### PR TITLE
feat(QA): Jira Ticket Automation

### PR DESCRIPTION
**Pull request by nshiland**
# Overview
Completely overhauled the existing ABR_robot_error script for expanded usage and QA applicability. Added better readability in attached files as well as title and description. Please refer to https://opentrons.atlassian.net/wiki/spaces/QT/pages/6215336005/How+to+setup+automated+Jira+ticketing for setup instructions.

Example Video: https://www.loom.com/share/7c153841ff724145a4f6d0a2adcea783

## Test Plan and Hands on Testing
Extensive use in QA and testing with various edge cases (SSH/HTTP inaccessible, etc). Ideally multiple QA users can set this up on their devices.

## Changelog
Made extensive changes to the ABR-testing/ABR_testing/data_collection folder, specifically read_robot_logs and robotFleet_error. Other small changes were made to scripts for error recovery and lint issues. Changed the makefile in ABR-testing to setup the script when prompted. 

As robotFleet_error is a new file, please look at effective changes from abr_robot_error in this link https://www.diffchecker.com/W3lIwQAX

## Review requests

- Please attempt to setup and use the script. Any feedback on the setup doc or script would be greatly appreciated.
- Please also look through the code. It is likely very messy. It (barely) passed lint, any comments on ways to go about things better would be greatly appreciated.

## Risk assessment

- To me this is overall very low risk due to it being confined to the ABR folders which are not used outside of QA/ABR.
- Medium risk to affecting existing ABR scripts, however, these were not widely used.

### CODE REVIEW

File: abr-testing/abr_testing/data_collection/read_robot_logs.py [hunk 1]
```diff
@@ -808,9 +851,10 @@ def save_run_log_to_json(
     ip: str, results: Dict[str, Any], storage_directory: Path
 ) -> str:
     """Save run log to local json file."""
-    data_file_name = ip + "_" + results["run_id"] + ".json"
+    data_file_name = "run_log.json"
```
Comment (y3rsh) [inline abr-testing/abr_testing/data_collection/read_robot_logs.py:854]: Does this not need a unique name anymore?

File: abr-testing/abr_testing/data_collection/robot_fleet_error.py
Comment (y3rsh) [inline abr-testing/abr_testing/data_collection/robot_fleet_error.py:1]: Please keep python filename like `robot_fleet_error`

File: abr-testing/abr_testing/data_collection/robot_fleet_error.py [hunk 2]
```diff
@@ -0,0 +1,965 @@
...
+        description,
+        project_key,
+        "-1",
+        "Bug",
+        "Medium",
+        data.components,
+        data.affects_version,
+        data.labels,
+        data.parent,
+    )
```
Comment (y3rsh) [inline abr-testing/abr_testing/data_collection/robot_fleet_error.py:974]: Please have these as named parameters instead of positional, it makes things clearer.

File: abr-testing/abr_testing/data_collection/robotFleet_error.py [hunk 3]
```diff
@@ -0,0 +1,965 @@
...
+        nargs=1,
+        help="Path to long term storage directory for run logs.",
+    )
+    args = parser.parse_args()
+    storage_directory = args.storage_directory[0]
+    credentials_path = os.path.join(storage_directory, "jiraCredentials.json")
+    with open(credentials_path) as f:
+        jiraCreds = json.load(f)
+    email = jiraCreds["Jira API"]["email"]
+    api_token = jiraCreds["Jira API"]["key"]
```
Comment (y3rsh) [inline abr-testing/abr_testing/data_collection/robotFleet_error.py]: It is desirable for this main function to be a clean sequence of calls to functions that resolve data and take action based on the arguments passed.  I want to be able to read this function and understand what is happening, not how.  

So, for example, abstracting these lines into a function in jira_tool that then return a  dataclass
would be an improvement.
```py
jira_credentials = jira_tool.get_credentials()
# then when you need the email
print(jira_credentials.email)
```

File: abr-testing/abr_testing/data_collection/robot_fleet_error.py [hunk 4]
```diff
@@ -0,0 +1,965 @@
...
+    args = parser.parse_args()
+    storage_directory = args.storage_directory[0]
+    credentials_path = os.path.join(storage_directory, "jiraCredentials.json")
+    with open(credentials_path) as f:
+        jiraCreds = json.load(f)
+    email = jiraCreds["Jira API"]["email"]
+    api_token = jiraCreds["Jira API"]["key"]
+    # Simplify the main section
+    board_id, project_key, ip, run_or_other = init_ticketing()
+    url = "https://opentrons.atlassian.net"
```
Comment (y3rsh) [inline abr-testing/abr_testing/data_collection/robot_fleet_error.py]: We have a lot of variables.  What I am wondering is if we can organize them into some groups.
There are things that may change, but usually not each run of the program.  These would be nice to resolve into a configuration dataclass we instantiate here near the top of main and then is passed around to all the functions that need it.

Jira, local machine storage paths, and the robot under target also feel like good candidates to gather into dataclasses to pass into the functions that need those resolved data.

File: abr-testing/abr_testing/data_collection/robot_fleet_error.py [hunk 5]
```diff
@@ -0,0 +1,965 @@
...
+    log_zip_path = read_robot_logs.get_logs(Path(storage_directory), ip)
+    ticket = jira_tool.JiraTicket(url, api_token, email)
+    try:
+        error_runs, protocol_ids = get_error_runs_from_robot(ip)
+    except requests.exceptions.InvalidURL:
+        print("Invalid IP address.")
+        sys.exit()
+    data = organize_ticket_data(
+        run_or_other, ip, storage_directory, project_key, error_runs, protocol_ids
+    )
```
Comment (y3rsh) [inline abr-testing/abr_testing/data_collection/robot_fleet_error.py:944]: when I look at this method signature I find it difficult to understand what data will be like.

File: abr-testing/abr_testing/data_collection/robotFleet_error.py [hunk 6]
```diff
@@ -0,0 +1,965 @@
...
+        run_or_other, ip, storage_directory, project_key, error_runs, protocol_ids
+    )
+
+    # make description replacement file
+    status_path = make_json_file(storage_directory, data.whole_description_str)
+    print(f"Making ticket for {data.summary}.")
+    all_issues = ticket.issues_on_board(project_key)
+    # CREATE TICKET
+    # TODO: for pyro, add pyro filter as HIGH priority
+    description = "Error recreation steps: (PLEASE FILL)"
```
Comment (y3rsh) [inline abr-testing/abr_testing/data_collection/robotFleet_error.py]: may all be pushed into create ticket?

File: abr-testing/abr_testing/data_collection/robotFleet_error.py [hunk 7]
```diff
@@ -0,0 +1,965 @@
...
+        "-1",
+        "Bug",
+        "Medium",
+        data.components,
+        data.affects_version,
+        data.labels,
+        data.parent,
+    )
+    issue_key = issue_key or ""
+    raw_issue_url = raw_issue_url or ""
```
Comment (y3rsh) [inline abr-testing/abr_testing/data_collection/robotFleet_error.py]: It feels weird here that these may be ""  
I am not tracking from the way things are organized and named why we would be proceeding in main if ticket creation failed.
If there is a fork here where there are no error logs, then make that explicit.

**Review by y3rsh — CHANGES_REQUESTED** [review]
Awesome work.  I think with some more thinking about the workflow as a whole, you can bring some organization to the code that will help future us follow what is going on more clearly.

File: abr-testing/abr_testing/data_collection/read_robot_logs.py (hunk 1 above)
Comment (nshiland) [reply abr-testing/abr_testing/data_collection/read_robot_logs.py:854]: Not anymore as I put them under the logs zip file which already has robot name and version. https://opentrons.atlassian.net/browse/RQA-5615 is a good example, making it just run_log.json made it a lot more readable. If you would like it put back the way it was, it is an easy fix.

File: abr-testing/abr_testing/data_collection/robot_fleet_error.py
Comment (nshiland) [reply abr-testing/abr_testing/data_collection/robot_fleet_error.py:1]: 100%, will update next week.

File: abr-testing/abr_testing/data_collection/robot_fleet_error.py (hunk 2 above)
Comment (nshiland) [reply abr-testing/abr_testing/data_collection/robot_fleet_error.py:974]: 100% for this. would you recommend doing this more broadly as well as it seems to be good practice?

File: abr-testing/abr_testing/data_collection/robotFleet_error.py (hunk 3 above)
Comment (nshiland) [reply abr-testing/abr_testing/data_collection/robotFleet_error.py]: Just put up a commit that *should* address this for the most part. Please let me know if there are other examples you think could be abstracted.

File: abr-testing/abr_testing/data_collection/robot_fleet_error.py (hunk 5 above)
Comment (nshiland) [reply abr-testing/abr_testing/data_collection/robot_fleet_error.py:944]: I assume this is talking about the function description. Yes, I just added a commit with this example and am now looking through all my function descriptions.

File: abr-testing/abr_testing/data_collection/robotFleet_error.py (hunk 7 above)
Comment (nshiland) [reply abr-testing/abr_testing/data_collection/robotFleet_error.py]: fixed. this made no sense and was a very poor fix by me to make lint happy

**Review by y3rsh — APPROVED** [review]
I like the improvements and I think it is good enough to merge and then iterate on as the edges emerge with actual use.

File: abr-testing/abr_testing/data_collection/robot_fleet_error.py (hunk 4 above)
Comment (nshiland) [reply abr-testing/abr_testing/data_collection/robot_fleet_error.py]: Going to leave this up, I did not address this. Will look at adding as a feature in the future.
